### DISCUSSION TITLE
retirement planning from current_age to life_expectancy_age without creating current_age - life_expectancy-age yearly 1-in subdirectories

### DISCUSSION THREAD

**Discussion by lestephane** [apauley/hledger-flow#105] [General]
I've tried a few retirement planners, and was not happy with any of them.
Excel is not something I look forward to using either, because it is not version control friendly.

So I'm looking to use hledger-flow, but the problem is that, if I need to model
all the expected cash flows from investments from this year (2021) to my year
of death (2070+ ,hopefully!), I would have to create, today, 50+ yearly subdirectories.

I can play games with symlinks to share as many input journals as possible, but I
would still end  up with 50+ yearly subdirectories.

The closest I think I could go is to create a script (yet another one) that I call manually
once in a while, that generates `1-in` directories from another input (say, `0-gen`) that contains
higher level definitions of the cash flows, something like `investmenta.gen` that contains:

```
InvestmentA-Account-Name
2021,-investmenta_inflow,InvestmentA-Inflow-Account-Name
2025,investmenta_outflow_1,InvestmentA-Outflow-Account-Name
2026,investmenta_outflow_2 # <-- no account name specified, uses last seen account name -> InvestmentA:Outflow-Account-Name 
yearly 2027-2040,investment_outflow_3_and_beyond #ditto
```

The output would be

```
1-in/2021
  investmenta.journal

    ~ 2021; forecast:
      InvestmentA-Inflow-Account-Name  -investmenta_inflow_y0
      InvestmentA-Account-Name

1-in/2022
  investmenta.journal (empty)
1-in/2023
  investmenta.journal (empty)
1-in/2024
  investmenta.journal (empty)
1-in/2025
  investmenta.journal

    ~ 2025; forecast:
      InvestmentA-Outflow-Account-Name  investmenta_outflow_1
      InvestmentA-Account-Name

1-in/2026
  investmenta.journal

    ~ 2026; forecast:
      InvestmentA:Outflow-Account-Name  investmenta_outflow_2
      InvestmentA-Account-Name

1-in/2027
  investmenta.journal

    ~ yearly 2027-2040; forecast:
      InvestmentA:Outflow-Account-Name  investment_outflow_3_and_beyond
      InvestmentA-Account-Name

1-in/2028
  investmenta.journal
  #Same as 2027
...
1-in/2040
  investmenta.journal
  #Same as 2027

 

```

I did not find anything related to retirement planning in the hledger issue tracker. Maybe the general retirment planning discussion will have to move there.

But the problem of needing one sub-directory per year, no matter what, is what I'm hoping to discuss here. I don't have a good solution. Can you think of a better way?

**Comment by apauley** [comment]
Yes, `hledger-flow` mostly assumes we're dealing with current and past transactions, the future is not really a well-defined concept.

What if we had a global `pre-import.journal` and `post-import.journal` that's not year-specific, and which would then be included in `import/all-years.journal`? It wouldn't be retirement-specific, which is probably better in any case.

And a meta question - what do you gain by combining your future planning with your past transactions? Does it allow you to have better reports?

I'm interested in knowing more about using hledger for retirement planning - please add links to any discussions or articles you come across.

**Reply by lestephane** [reply]
I did not find any discussion on retirement planning in relation to command-line accounting, so I can't link to anything at this point.

The reason I keep track of everything in the same place is because of lazyness. 2 businesses, 1 foundation, 1 individual, and the transfer transactions between them are easier to track if it's all in the same journal tree.

It's also important for me to have at all times the past and the provisional future, where I can ignore future transactions if I only care about the current state of the world. I do not use periodic transactions beyond `1-in`, all those periodic transactions get 'materialized' with an additional `forecast:` tag in the `3-journal` directory, and I can exclude future transactions using not:tag:forecast.

I also keep timeclock files in the same filesystem tree. It will eventually take too much time to hledger-flow import everything, at which point I'll try to segregate, but I'm not looking at that just yet. When I do, maybe the idea of having a top-level (pre|post)-import will make sense.

![Selection_001](https://user-images.githubusercontent.com/5460884/133734955-49e2e148-821c-44a4-bbfb-aae6f7e7ab69.png)

This is what I've got so far:

## generate.inputs
```
Foundation-XX:Current-Account:USD
every December 1st@2032-2077,10000,Personal:Assets:XX:USD
```
## import/foundation-xx/distributions/primary-beneficiary/1-in/2032/primary-beneficiary-distributions-2032.journal

```
~ every December 1st 2032  * "primary-beneficiary distributions outflow"; forecast:, type:cashflow
  Personal:Assets:XX:USD  10000 USD
  Foundation-XX:Current-Account:USD
```

## import/foundation-xx/distributions/primary-beneficiary/3-journal/2032/primary-beneficiary-distributions-2032.journal

```
2032-12-01 * "primary-beneficiary distributions outflow"  ; forecast:, type:cashflow
    ; generated-transaction: ~ every December 1st 2032
    Personal:Assets:XX:USD             10000 USD
    Foundation-XX:Current-Account:USD
```
## ...Ditto until 2077


If you're interested I can send the bash script that generates the inputs. But it's not finished yet, I'm already having to add a capability to increase or decrease a given cashflow by X% per period. To model for example a yearly 4% increase in international health insurance premiums.
